addons/pos_loyalty/models/pos_session.py

```diff
@@ -0,0 +1,94 @@
+            result.append('loyalty.reward')
+        return result
+
+    def _loader_params_loyalty_program(self):
+        return {
+            'search_params': {
+                'domain': [('id', 'in', self.config_id.all_program_ids.ids), ('active', '=', True)],
+                'fields': ['name', 'trigger', 'applies_on', 'program_type', 'date_to', 'limit_usage', 'max_usage', 'is_nominative'],
+            },
+        }
+
+    def _loader_params_loyalty_rule(self):
+        return {
+            'search_params': {
+                'domain': [('program_id', 'in', self.config_id.all_program_ids.ids)],
+                'fields': ['program_id', 'valid_product_ids', 'any_product', 'currency_id',
+                    'reward_point_amount', 'reward_point_split', 'reward_point_mode',
+                    'minimum_qty', 'minimum_amount', 'minimum_amount_tax_mode', 'mode', 'code'],
+            }
+        }
+
+    def _loader_params_loyalty_reward(self):
+        return {
+            'search_params': {
+                'domain': [('program_id', 'in', self.config_id.all_program_ids.ids)],
+                'fields': ['description', 'program_id', 'reward_type', 'required_points', 'clear_wallet', 'currency_id',
+                    'discount', 'discount_mode', 'discount_applicability', 'all_discount_product_ids', 'is_global_discount',
+                    'discount_max_amount', 'discount_line_product_id',
+                    'multi_product', 'reward_product_ids', 'reward_product_qty', 'reward_product_uom_id'],
+            }
+        }
+
+    def _get_pos_ui_loyalty_program(self, params):
+        return self.env['loyalty.program'].search_read(**params['search_params'])
+
+    def _get_pos_ui_loyalty_rule(self, params):
+        return self.env['loyalty.rule'].search_read(**params['search_params'])
+
+    def _get_pos_ui_loyalty_reward(self, params):
+        return self.env['loyalty.reward'].search_read(**params['search_params'])
+
+    def _get_pos_ui_res_partner(self, params):
+        result = super()._get_pos_ui_res_partner(params)
+        # In order to make loyalty programs work offline we load the partner's point into
+        # a non-existant field 'loyalty_points'.
+        if self.config_id.loyalty_program_id:
+            # collect ids in a list, group by id and default points to 0
+            partner_ids = []
+            res_by_id = {}
+            for res in result:
+                partner_ids.append(res['id'])
+                res_by_id[res['id']] = res
+                res['loyalty_points'] = 0
+                res['loyalty_card_id'] = False
+            # Direct query to avoid loading loyalty cards in the cache for no reason.
+            # There is no context where we would need to flush.
+            query = self.env['loyalty.card']._where_calc(
+                [('program_id', '=', self.config_id.loyalty_program_id.id), ('partner_id', 'in', partner_ids)]
+            )
+            self.env['loyalty.card']._apply_ir_rules(query)
```

Comment:
I think you could use ``_search``, which does the same thing and returns a query.